### ISSUE TITLE
PEP420 `cyclic-import` false positive

### ISSUE THREAD

**Issue by alexey-pelykh** [PyCQA/pylint#7959]
### Bug description

https://github.com/alexey-pelykh/reproduce-pylint-pep420-cyclic-import-false-positive

`pylint` thinks that `logging` is the package name while due to implicit namespace, it's `some.package.logging`.

### Configuration

```ini
[MASTER]
init-hook=
    from os.path import dirname, join
    from glob import glob
    from pylint.config import find_pylintrc
    from site import addsitedir
    site_packages_paths = glob(join(dirname(find_pylintrc()), '**', '*venv', '**', 'site-packages'), recursive=True)
    for site_packages_path in site_packages_paths: addsitedir(site_packages_path)
```

### Command used

```shell
./venv/bin/pylint --recursive=y src
./venv/bin/pylint $(find src -name '*.py')
```

### Pylint output

```shell
************* Module logging.wrapper.__init__
src/some/package/logging/wrapper/__init__.py:1:0: R0401: Cyclic import (logging -> logging.wrapper) (cyclic-import)
```

### Expected behavior

No cyclic import detected

### Pylint version

```shell
pylint 2.15.9
astroid 2.12.13
Python 3.11.0 (main, Oct 26 2022, 19:06:18) [Clang 14.0.0 (clang-1400.0.29.202)]
```

**Comment by alexey-pelykh** [comment]
@Pierre-Sassoulas the reproduction repo linked did not work as reproduction for you?

**Comment by Pierre-Sassoulas** [comment]
I did not have the time to try to reproduce, sorry about that. Multiple files reproducer and full repo reproducer takes longer to label and triage, but I usually reproduce when triaging and before labelling so the 'needs reproduction' label is never applied for most issue.

**Comment by alexey-pelykh** [comment]
@Pierre-Sassoulas I've looked a bit deeper into the origins of the issue, it's more or less clear that with `--recursive=y` there's no way the implicit namespace will be detected since `pylint` no "sources root(s)" option to define from where to detect these implicit namespaces. That said, any insights on "source root(s)" concept in `pylint`? Discussions, PRs, etc?

**Comment by alexey-pelykh** [comment]
A side note, that my original use case is with `pre-commit` where set of files is passed to `pylint` - which also disallows `pylint` to figure out the proper package name since no `__init__.py` in implicit namespaces, therefore concept of "source root(s)" pops up again.

**Comment by Pierre-Sassoulas** [comment]
Thank you for looking into this ! 

> That said, any insights on "source root(s)" concept in pylint? Discussions, PRs, etc?

That could be the huge merge request where the ``--recursive`` option was introduced (https://github.com/PyCQA/pylint/pull/5682) and possibly subsequent correcting MR, but I don't think we talked about the concept of source root here. Might be a hole in the design, but I think we mainly tried to implement the equivalent of the bash command "find all python file in this directory" that was used a lot in continuous integration (``find . -iname "*.py" -exec pylint -E {} ;\`` & co.).

**Comment by alexey-pelykh** [comment]
Thanks, I'll look into the discussion! Yet from what I can see from the sources thus far, a PR is needed to introduce source root(s) concept to the configuration files to produce package names relative to those roots when the file is discovered under those roots while doing nothing to the package names if it's outside. The default list of source roots would be empty, not affecting the default behavior while fixing the issue for me :)
